src/cargo/util/toml_mut/manifest.rs

```diff
@@ -267,18 +270,56 @@ impl LocalManifest {
         if !path.is_absolute() {
             anyhow::bail!("can only edit absolute paths, got {}", path.display());
         }
-        let data = cargo_util::paths::read(&path)?;
+        let raw = cargo_util::paths::read(&path)?;
+        let mut data = raw.clone();
+        let mut embedded = None;
+        if is_embedded(path) {
+            let source = ScriptSource::parse(&data)?;
+            if let Some(frontmatter) = source.frontmatter() {
+                embedded = Some(Embedded::exists(&data, frontmatter));
+                data = frontmatter.to_owned();
+            } else if let Some(shebang) = source.shebang() {
+                embedded = Some(Embedded::after(&data, shebang));
+                data = String::new();
+            } else {
+                embedded = Some(Embedded::start());
+                data = String::new();
+            }
+        }
         let manifest = data.parse().context("Unable to parse Cargo.toml")?;
         Ok(LocalManifest {
             manifest,
             path: path.to_owned(),
+            raw,
+            embedded,
         })
     }
 
     /// Write changes back to the file.
     pub fn write(&self) -> CargoResult<()> {
-        let s = self.manifest.data.to_string();
-        let new_contents_bytes = s.as_bytes();
+        let mut manifest = self.manifest.data.to_string();
+        let raw = match self.embedded.as_ref() {
+            Some(Embedded::Implicit(start)) => {
+                if !manifest.ends_with("\n") {
+                    manifest.push_str("\n");
+                }
+                let fence = "---\n";
+                let prefix = &self.raw[0..*start];
+                let suffix = &self.raw[*start..];
```

Comment:
Only if you want, either form seems fine:

```suggestion
                let (prefix, suffix) = self.raw.split_at(*start);
```